### PR TITLE
scheduled caching of metrics data

### PR DESCRIPTION
**Pull request by shah-ee**
# Context

- When testing in labs env, turned out there were a lot of entries being returned and since for each dependency, we need to look up information about them on teams and repositories, there would be N look ups, if /api/dependencies were to return N entries. Even though, we would cache the results, this was leading to time-outs as the info was being looked up at the time of the lookup.
- Sadly filtering out archived repos, or libraries can only be done after we have hit teams-and-repos endpoint, so the filtering doesn't reduce the number of remote calls.
- This PR sets up schedulers to do this asynchronously to populate the cache entries.
- It introduces some generic types to make the potential debugging and maintenance clearer.

# Observed local behaviour

- the [endpoint](http://localhost:9023/platform-progress-metrics/metrics) returns an empty list to start with. after the initial delay however, it populates the data and returns the full results.

# Todo
- [ ] It's probably not a good idea to rely on the injected `ExecutionContext` and best specify different ones for the different entities:
  - client for /api/dependencies call
  - client for /teams-and-repos call
  - scheduler for populating api dependencies
  - scheduler for populating teams and repos data
- [ ] remove scheduled caching in favor of top level call to teams-and-repos to filter out libraries and deprecated services. 

# Question

- How often should the schedulers update the data? (current hard-coded data need to change as i have kept them low to test the behaviour locally)
- Could do with better names for `Cache0`, `Cache1`, `Fetch0` and `Fetch1` :)

### CODE REVIEW

File: app/uk/gov/hmrc/platformprogressmetrics/connectors/Fetcher0.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  @Singleton
+  class OfAllRepositories @Inject()(dependenciesCache: Cache0[List[RepoDependencies]])(implicit val ec: ExecutionContext) extends Fetcher0[List[Repository]]{
+    override def fetch: Future[List[Repository]] = dependenciesCache.get.map {
+      case Some(repos) => repos.map(_.repositoryName)
+      case None => Nil
+    }
+  }
+
+}
```
Comment (shah-ee) [inline app/uk/gov/hmrc/platformprogressmetrics/connectors/Fetcher0.scala:64]: add eof.

File: app/uk/gov/hmrc/platformprogressmetrics/connectors/Fetcher1.scala [hunk 2]
```diff
@@ -76,11 +51,7 @@ object TeamsAndRepositoriesConnector{
...
         }
-        .flatTap(cache.set(repository))
     }
-
-    override def fetch(repository: Repository): Future[List[Team]] = OptionT(cache.get(repository))
-      .getOrElseF(refresh(repository))
   }
 
-}
+}
```
Comment (shah-ee) [inline app/uk/gov/hmrc/platformprogressmetrics/connectors/Fetcher1.scala:57]: add eof

File: app/uk/gov/hmrc/platformprogressmetrics/schedulers/ScheduledCache0.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  //ugly hack till I learn how to bind type literals properly.
+  @Singleton
+  class OfRepoDependencies @Inject()(actorSystem: ActorSystem, underlyingCache: Cache0[List[RepoDependencies]], fetcher: Fetcher0[List[RepoDependencies]])(
+    implicit executionContext: ExecutionContext
+  ) extends Impl(actorSystem, underlyingCache, fetcher)
+
+
+}
```
Comment (shah-ee) [inline app/uk/gov/hmrc/platformprogressmetrics/schedulers/ScheduledCache0.scala:57]: eof.

File: app/uk/gov/hmrc/platformprogressmetrics/schedulers/ScheduledCache1.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    override def fetch(key: K): Future[Option[V]] = underlyingCache.get(key)
+
+  }
+
+  class OfTeamsAndRepos @Inject() (actorSystem: ActorSystem, underlyingCache: Cache1[Repository, List[Team]], fetchByKey: Fetcher1[Repository, List[Team]], fetchAll: Fetcher0[List[Repository]])(implicit executionContext: ExecutionContext) extends ScheduledCache1.Impl[Repository, List[Team]](
+    actorSystem, underlyingCache, fetchByKey, fetchAll
+  )
+
+}
```
Comment (shah-ee) [inline app/uk/gov/hmrc/platformprogressmetrics/schedulers/ScheduledCache1.scala:59]: eof

**Review by colin-lamed — COMMENTED** [review]
I'm not convinced the generic types and abstractions make the logic clearer or easier to maintain - it's a small application so not sure we need the excessive levels of indirection (ala J2EE).
Wrapping the API calls with `play.api.cache.AsyncCacheApi.getOrElseUpdate(key, cacheExpiry)` would be a lot simpler - i.e on demand (no empty resuts). Another pattern we have used across our services is to use mongo to cache data across all instances.

Why do we need N calls to team-and-repositories? A single call to `/api/repositories?archived=false` should be enough to get a list of non-archived repositories to filter the results of service-dependencies `/api/dependencies`?

**Comment by shah-ee** [comment]
Thanks @colin-lamed 

> Why do we need N calls to team-and-repositories? A single call to /api/repositories?archived=false should be enough to get a list of non-archived repositories to filter the results of service-dependencies /api/dependencies?

- ah very useful thanks! didn't know we could filter the results [like that](https://github.com/hmrc/teams-and-repositories/blob/master/conf/app.routes#L11). we don't seem to have the equivalent for `type=library` tho. so I think the point stands.

> I'm not convinced the generic types and abstractions make the logic clearer or easier to maintain - it's a small application so not sure we need the excessive levels of indirection

- So there is two points here:
  - abstractions
  - generic types

- as for the abstractions, we have three of them: `Fetch`, `Cache` and `ScheduledCache`. (ignoring the difference in flavour for the minute, ie 0 vs 1 versions) 
 
The actual logical steps are:
- fetch the data
- cache the data
- have a mechanism to get the data asynchronously

While `AsyncCacheApi` is a good idea, the existence of a `Cahce` interface (maybe I should call it `Store`?) allows us to swap a Caffine impl with a Mongo one without changing the code too much. so i am not sure it's excessive. What do we think? re: the J2EE reference, only one call is involved when you hit `/metrics` so not sure there is a performance issue here.

- as for generic types, well previously the act of fetching and caching the data were tied together in the individual clients. this separates it and allows for easier reasoning about the code. also yes it is a small application, but the different metrics will have different impl but will all follow the same pattern so I think the generic types are serving a purpose.

**Pull request closed without merging**